src/AggregateFunctions/SingleValueData.cpp

```diff
@@ -0,0 +1,1419 @@
+
+StringRef SingleValueDataString::getStringRef() const
+{
+    return StringRef(getData(), size);
+}
+
+void SingleValueDataString::allocateLargeDataIfNeeded(UInt32 size_to_reserve, Arena * arena)
+{
+    if (capacity < size_to_reserve)
+    {
+        if (unlikely(MAX_STRING_SIZE < size_to_reserve))
+            throw Exception(
+                ErrorCodes::TOO_LARGE_STRING_SIZE, "String size is too big ({}), maximum: {}", size_to_reserve, MAX_STRING_SIZE);
+
+        size_t rounded_capacity = roundUpToPowerOfTwoOrZero(size_to_reserve);
+        chassert(rounded_capacity <= MAX_STRING_SIZE + 1); /// rounded_capacity <= 2^31
+        capacity = static_cast<UInt32>(rounded_capacity);
+
+        /// Don't free large_data here.
+        large_data = arena->alloc(capacity);
+    }
+}
+
+void SingleValueDataString::changeImpl(StringRef value, Arena * arena)
+{
+    if (unlikely(MAX_STRING_SIZE < value.size))
+        throw Exception(ErrorCodes::TOO_LARGE_STRING_SIZE, "String size is too big ({}), maximum: {}", value.size, MAX_STRING_SIZE);
+
+    UInt32 value_size = static_cast<UInt32>(value.size);
+
+    if (value_size <= MAX_SMALL_STRING_SIZE)
+    {
+        /// Don't free large_data here.
+        size = value_size;
+
+        if (size > 0)
+            memcpy(small_data, value.data, size);
+    }
+    else
+    {
+        allocateLargeDataIfNeeded(value_size, arena);
+        size = value_size;
+        memcpy(large_data, value.data, size);
+    }
+}
+
+void SingleValueDataString::insertResultInto(DB::IColumn & to) const
+{
+    if (has())
+        StringValueCompatibility::insertDataWithTerminatingZero(assert_cast<ColumnString &>(to), getData(), size);
+    else
+        assert_cast<ColumnString &>(to).insertDefault();
+}
+
+void SingleValueDataString::write(WriteBuffer & buf, const ISerialization & /*serialization*/) const
+{
+    if (unlikely(MAX_STRING_SIZE < size))
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "String size is too big ({}), it's a bug", size);
+
+    /// For serialization we use signed Int32 (for historical reasons), -1 means "no value"
+    Int32 size_to_write = size ? size : -1;
+    writeBinaryLittleEndian(size_to_write, buf);
+    if (has())
+        buf.write(getData(), size);
+}
+
+void SingleValueDataString::read(ReadBuffer & buf, const ISerialization & /*serialization*/, Arena * arena)
+{
+    /// For serialization we use signed Int32 (for historical reasons), -1 means "no value"
+    Int32 rhs_size_signed;
+    readBinaryLittleEndian(rhs_size_signed, buf);
+
+    if (rhs_size_signed < 0)
+    {
+        /// Don't free large_data here.
+        size = 0;
+        return;
+    }
+
+    UInt32 rhs_size = rhs_size_signed;
+    if (rhs_size <= MAX_SMALL_STRING_SIZE)
+    {
+        /// Don't free large_data here.
+        size = rhs_size;
+        buf.readStrict(small_data, size);
+    }
+    else
+    {
+        /// Reserve one byte more for null-character
+        allocateLargeDataIfNeeded(rhs_size + 1, arena);
```

Comment:
I'm bit confused, doesn't size already take into account `null-character`

Reply:
This is old (unchanged) code. For some releases size didn't include the null character, so now we pay for it always allocating an extra byte just in case we need it. More context in https://github.com/ClickHouse/ClickHouse/pull/43038